### PR TITLE
Fixed issues

### PR DESCRIPTION
**Pull request by FeitianSmartcardReader**
1. Solved multiple epss2003 usage issue
2. check expats point to prevent memory leak
3. Add new ATR for entersafe PKI card
4. declare all variables at the beginning of block
5. Solved Incorrect PIN raise wrong CKR error, no token flags change

### CODE REVIEW

**Comment by FeitianSmartcardReader** [comment]
Solved #871 #731 #730

File: src/libopensc/card-epass2003.c [hunk 1]
```diff
@@ -106,11 +107,100 @@ typedef struct epass2003_exdata_st {
 		((unsigned long)x & 0x0000FF00)<<  8	| \
 		((unsigned long)x & 0x000000FF)<< 24)
 
+#define pin_low_notify(ret) \
+      printf("Verification failed (remaining tries: %d)\n",ret)
```
Comment (frankmorgner) [inline src/libopensc/card-epass2003.c]: indenting: use tabs instead of spaces

File: src/libopensc/card-epass2003.c [hunk 2]
```diff
@@ -106,11 +107,100 @@ typedef struct epass2003_exdata_st {
...
+
+static const struct sc_card_error epass2003_errors[] = {
+	{ 0x6200, SC_ERROR_CARD_CMD_FAILED,	"Warning: no information given, non-volatile memory is unchanged" },
+	{ 0x6281, SC_ERROR_CORRUPTED_DATA,	"Part of returned data may be corrupted" },
+	{ 0x6282, SC_ERROR_FILE_END_REACHED,	"End of file/record reached before reading Le bytes" },
+    { 0x6283, SC_ERROR_CARD_CMD_FAILED,	"Selected file invalidated" },
```
Comment (frankmorgner) [inline src/libopensc/card-epass2003.c]: indenting: use tabs instead of spaces

File: src/libopensc/card-epass2003.c [hunk 3]
```diff
@@ -106,11 +107,100 @@ typedef struct epass2003_exdata_st {
...
 int epass2003_refresh(struct sc_card *card);
 
 static int
+epass2003_check_sw(struct sc_card *card, unsigned int sw1, unsigned int sw2)
+{
+	const int err_count = sizeof(epass2003_errors)/sizeof(epass2003_errors[0]);
+	int i;
+    
+	/* Handle special cases here */
+    if (sw1 == 0x6C) {
```
Comment (frankmorgner) [inline src/libopensc/card-epass2003.c]: indenting: use tabs instead of spaces

File: src/libopensc/card-epass2003.c (hunk 1 above)
Comment (frankmorgner) [inline src/libopensc/card-epass2003.c]: don't use `printf`. instead use `sc_log`!

**Comment by frankmorgner** [comment]
@velter , @ValenMX, could you please check if this fixes your issues?

**Comment by FeitianSmartcardReader** [comment]
@frankmorgner  the code was confirmed by ValenVM and Velter in https://github.com/OpenSC/OpenSC/pull/787 and https://github.com/OpenSC/OpenSC/pull/770

File: src/libopensc/card-epass2003.c [hunk 4]
```diff
@@ -179,25 +177,25 @@ int epass2003_refresh(struct sc_card *card);
...
 epass2003_check_sw(struct sc_card *card, unsigned int sw1, unsigned int sw2)
 {
-	const int err_count = sizeof(epass2003_errors)/sizeof(epass2003_errors[0]);
-	int i;
-    
-	/* Handle special cases here */
+    const int err_count = sizeof(epass2003_errors)/sizeof(epass2003_errors[0]);
+    int i;
+
+    /* Handle special cases here */
```
Comment (frankmorgner) [inline src/libopensc/card-epass2003.c]: Unfortunately, you commit introduces new spaces...

File: src/libopensc/card-epass2003.c [hunk 5]
```diff
@@ -2493,14 +2493,15 @@ update_secret_key(struct sc_card *card, unsigned char ktype, unsigned char kid,
 	return r;
 }
 
+#if 1
```
Comment (frankmorgner) [inline src/libopensc/card-epass2003.c]: should be removed

**Comment by FeitianSmartcardReader** [comment]
I see, will check it, thanks 
 


Best Regard 
 
Ben |Technical Specialist (GLBAL SUPPORT) | 国际部 王红宾
Skype: Hooben@FEITIAN | Tel: +86 (10) 62304466 ext. 575
Feitian, Inc. Floor 17th, Tower B, Huizhi Mansion, No.9 Xueqing Road, Haidian District Beijing 100085, P.R China
Smartcard Readers-> http://www.ftsafe.com/products/reader
Online Shop->http://ftsafe.com/onlinestore/index
 
This email message (including any attachments) is for the sole use of the intended recipient and may contain confidential and privileged information. Any unauthorized review, use, disclosure or distribution of this information is prohibited. If you are not the intended recipient, please contact the sender by reply email and destroy the original message and all of its copies.
 
From: Frank Morgner
Date: 2016-10-10 18:21
To: OpenSC/OpenSC
CC: Feitian Technologies; Author
Subject: Re: [OpenSC/OpenSC] Fixed issues (#879)
@frankmorgner commented on this pull request.


In src/libopensc/card-epass2003.c:
> @@ -179,25 +177,25 @@ int epass2003_refresh(struct sc_card *card);
 static int
 epass2003_check_sw(struct sc_card *card, unsigned int sw1, unsigned int sw2)
 {
-	const int err_count = sizeof(epass2003_errors)/sizeof(epass2003_errors[0]);
-	int i;
-    
-	/* Handle special cases here */
+    const int err_count = sizeof(epass2003_errors)/sizeof(epass2003_errors[0]);
+    int i;
+
+    /* Handle special cases here */

Unfortunately, you commit introduces new spaces...


In src/libopensc/card-epass2003.c:
> @@ -2493,14 +2493,15 @@ update_secret_key(struct sc_card *card, unsigned char ktype, unsigned char kid,
 	return r;
 }
 
+#if 1

should be removed

**Comment by frankmorgner** [comment]
I fixed the outstanding changes

**Comment by frankmorgner** [comment]
... and pushed them to master

**Comment by FeitianSmartcardReader** [comment]
@frankmorgner Thanks

**Pull request closed without merging**